Updates  to rocm recipes for rocm-5.5.0 and rocm-5.5.1 releases

srekolam opened the pull request:
Initial commit for rocm-5.5.0 release. Planning to update this PR with rocm-5.5.1 as it is released today.

srekolam committed:
initial commit for rocm-5.5.0 release

srekolam committed:
Merge branch 'develop' into rocm-5.5.0-rel-br1

srekolam committed:
fix style errors

cgmb commented on var/spack/repos/builtin/packages/hip/0014-remove-compiler-rt-linkage-for-host.5.5.0.patch:
```diff
@@ -0,0 +1,58 @@
+         }
+diff --git a/hipamd/hip-config.cmake.in b/hipamd/hip-config.cmake.in
+index 7c4fe7f..4f2dc19 100755
+--- a/hipamd/hip-config.cmake.in
++++ b/hipamd/hip-config.cmake.in
+@@ -253,7 +253,6 @@ if(HIP_COMPILER STREQUAL "clang")
+   # Add support for __fp16 and _Float16, explicitly link with compiler-rt
+   if( "${CLANGRT_BUILTINS_FETCH_EXIT_CODE}" STREQUAL "0" )
+     # CLANG_RT Builtins found Successfully Set interface link libraries property
+-    set_property(TARGET hip::host APPEND PROPERTY INTERFACE_LINK_LIBRARIES "${CLANGRT_BUILTINS}")
+     set_property(TARGET hip::device APPEND PROPERTY INTERFACE_LINK_LIBRARIES "${CLANGRT_BUILTINS}")
+   else()
+     message(STATUS "clangrt builtins lib not found: ${CLANGRT_BUILTINS_FETCH_EXIT_CODE}")
```
In the interest of eventually eliminating this patch, could we maybe take one thing from this patch and see if we can get it fixed upstream in hipamd itself? Of the things here, I think the most valuable would probably be getting `${CLANGRT_BUILTINS}` addressed.

srekolam replied:
sure.. i will check on this.

cgmb commented on var/spack/repos/builtin/packages/llvm-amdgpu/package.py:
```diff
@@ -163,6 +165,7 @@ class LlvmAmdgpu(CMakePackage):
     # as per 5.2.0 llvm code. It used to be llvm/bin/../lib/libdevice.
     # Below patch is to look in the old path.
     patch("adjust-openmp-bitcode-directory-for-llvm-link.patch", when="@5.2.0:")
+    patch("patch-llvm-5.5.0.patch", when="@5.5.0")
```
This should have a more descriptive filename and a comment explaining why it is needed.

srekolam replied:
basically, with llvm-amdgpu@5.5.0, it seems to support the code-object-version 5 but for some reason the changes in the hsa-rocr-dev for 5.5.0 in the file  https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/master/src/image/blit_src/CMakeLists.txt seems to break for spack build when building hsa-rocr-dev.  Rightnow, this workaround is put for this reason. Still checking if there is anything that can be done.

srekolam replied:
basically this line inside https://github.com/RadeonOpenCompute/ROCR-Runtime/blob/master/src/image/blit_src/CMakeLists.txt 

 separate_arguments(CLANG_ARG_LIST UNIX_COMMAND
    "-O2 -x cl -Xclang -finclude-default-header -cl-denorms-are-zero -cl-std=CL2.0
    -target amdgcn-amd-amdhsa -mcpu=${TARGET_ID} **-mcode-object-version=4**
    -o ${OUTPUT_FILE} ${INPUT_FILE}")

cgmb reviewed (commented):
The libraries look good. I had a couple comments on other packages, but nothing too important.

srekolam committed:
fix the hipsparse  build error for 5.5.0

srekolam committed:
fix build error for amrex .add hiprand as a dependency

srekolam committed:
modify the patch for rocprofiler-dev

srekolam committed:
add hiprand for +rocm build

srekolam committed:
initial commit for rocm-5.5.1 release

srekolam committed:
bump up the version for rocm-5.5.1 release.

srekolam committed:
bump up the version for rocmlir.miopen to use this backend only till
5.5

srekolam committed:
Merge branch 'develop' into rocm-5.5.0-rel-br1

srekolam committed:
add new recipe py-barectf and add it as dependency for rocprofiler-dev

srekolam commented:
@spackbot rerun pipeline

srekolam committed:
revert the changes for rocprofiler-dev for 5.5.0/1 for now as it depends
on hsa-amdaqlprofile.so which is a closed source and no spack recipe is
available for now.

srekolam committed:
add rocm-core as dependency for rocm packages from 5.5.0 onwards

srekolam committed:
fix style errors

srekolam committed:
fix style error

cgmb commented on var/spack/repos/builtin/packages/comgr/package.py:
```diff
@@ -157,6 +160,9 @@ class Comgr(CMakePackage):
             "rocm-device-libs@" + ver, when="@{0} ^llvm-amdgpu ~rocm-device-libs".format(ver)
         )
 
+    for ver in ["5.5.0", "5.5.1"]:
+        depends_on("rocm-core@" + ver, when="@" + ver)
```
Does it really depend on rocm-core? That would be very unfortunate.

srekolam replied:
yeah i think so.. you can refer here . https://github.com/RadeonOpenCompute/ROCm-CompilerSupport/blob/amd-stg-open/lib/comgr/CMakeLists.txt

cgmb replied:
In the deb/rpm packages, the declared dependencies on rocm-core exist only so that you can `apt remove rocm-core` or `dnf remove rocm-core` and the package manager will remove all other rocm components.

The new thing in ROCm 5.5 is that a rocmcore library was added to the rocm-core package. The library that can be used to query the version number of the rocm release. However, nothing should actually be using that library. The version APIs it provides only really make sense in the context of the AMD deb/rpm releases.

cgmb commented:
This is a bit of a rant, but the creation of the rocmcore library is very unfortunate. There is no way to correctly use the version number that it provides outside the official AMD RPM/DEB packages. For example, on Debian Bookworm, the version of HIP is 5.2.3 but most of the math libraries are 5.3.3. If users care about the version of HIP, they should query the version of HIP and if users care about the version of rocSPARSE, they should query the version of rocSPARSE. Querying rocmcore and assuming that all components match the rocmcore version will lead to nothing but headaches when that assumption is violated.

srekolam committed:
Merge branch 'develop' into rocm-5.5.0-rel-br1

srekolam committed:
Merge branch 'develop' into rocm-5.5.0-rel-br1

srekolam committed:
Merge branch 'develop' into rocm-5.5.0-rel-br1

srekolam committed:
Merge remote-tracking branch 'upstream/develop' into rocm-5.5.0-rel-br1

srekolam commented:
can i get some reviews and if things are okay, approval..pls.

srekolam commented:
@tldahlgren , can you please let me know if this can be merged?

srekolam committed:
Merge branch 'develop' into rocm-5.5.0-rel-br1

srekolam commented:
@spackbot rerun pipeline

srekolam commented:
@spackbot rerun pipeline

srekolam committed:
Merge branch 'develop' into rocm-5.5.0-rel-br1

srekolam committed:
avoid download of the gtest for building unit tests

srekolam commented:
@tldahlgren , can you please review this and let me know if this can be merged,

srekolam committed:
Merge branch 'develop' into rocm-5.5.0-rel-br1

tldahlgren reviewed (approved):
Re-confirmed all of the version sha256.

msimberg commented:
@srekolam sorry commenting way past this is merged, but I've been trying to get rocprof working through spack. What was the real reason for reverting the `rocprofiler-dev` changes inhttps://github.com/spack/spack/pull/37910/commits/82d45467ef5c9d29bcf3acec5cdc21859c0ef9da? `rocprofiler-dev` already has a patch for ignoring aqlprofile for older version: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/rocprofiler-dev/0001-Continue-build-in-absence-of-aql-profile-lib.patch. Did the aql stuff go from being easily ignorable to being more difficult to ignore? I've seen comment about it being closed source so it makes sense that one would want to remove that dependency. I'm mainly just wondering what happened in 5.5.0 regarding that?

srekolam commented:
basically libhsa-amd-aqlprofile64.so has been a closed source binary and due to which we could not get that in the spack build for rocprofiler.
The patches inside the rocprofiler-dev was trying to get past the build failures that were seen due to absence of this libhsa-amd-aqlprofile64.so . 
the initially patches were able to work around it but from 5.5.0 onwards there were changes to the rocprofiler-dev makefiles which made it more difficult to get past the failures.
if you have the libhsa-amd-aqlprofile64.so in the /opt/rocm-xx.xx.xx/lib path then i believe you would not need this patch.
There is also a plan to have the aqlprofile as a public git hub project but that is still under some discussion.
